scarb/src/compiler/compilers/test.rs

```diff
@@ -13,10 +17,44 @@ impl Compiler for TestCompiler {
 
     fn compile(
         &self,
-        _unit: CompilationUnit,
-        _db: &mut RootDatabase,
-        _ws: &Workspace<'_>,
+        unit: CompilationUnit,
+        db: &mut RootDatabase,
+        ws: &Workspace<'_>,
     ) -> Result<()> {
+        let target_dir = unit.target_dir(ws.config());
+
+        let test_crate_ids = collect_main_crate_ids(&unit, db);
+        let main_crate_ids = collect_all_crate_ids(&unit, db);
+        let starknet = unit.cairo_plugins.iter().any(|plugin| {
+            plugin.package.id.name == PackageName::STARKNET
+                && plugin.package.id.source_id == SourceId::for_std()
+        });
+
+        let diagnostics_reporter = DiagnosticsReporter::callback({
+            let config = ws.config();
+            |diagnostic: String| {
+                config
+                    .ui()
+                    .print(TypedMessage::naked_text("diagnostic", &diagnostic));
+            }
+        });
+
+        diagnostics_reporter
+            .with_extra_crates(&main_crate_ids)
+            .ensure(db)?;
+
+        let test_compilation =
+            compile_test_prepared_db(db, starknet, main_crate_ids, test_crate_ids)?;
```

Comment:
This is **not** going to scale. All Scarb changes to root database flow will require upstream changes in the compiler. We need to be able to end-to-end control what happens with the root database. Perhaps compiler APIs should focus on collecting test-specific information only?

Reply:
Discussed offline.